pkg/jobworker/worker.go

```diff
@@ -0,0 +1,219 @@
+package jobworker
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"sync"
+	"syscall"
+
+	"github.com/google/uuid"
+)
+
+// CgroupByte is used as a Byte to parse JobOpts.MemLimit cgroup value
+type CgroupByte int64
+
+func (b CgroupByte) String() string {
+	return fmt.Sprintf("%d", b)
+}
+
+const (
+	CgroupKB CgroupByte = 1024
+	CgroupMB            = CgroupKB * 1024
+	CgroupGB            = CgroupMB * 1024
+)
+
+// Job maintains the exec.Cmd struct (containing the underlying os.Process) and the "owner" for authz
+type Job struct {
+	sync.RWMutex
+	ID      string
+	running bool
+	done    chan bool
+	pgid    int
+	cmd     *exec.Cmd
+	con     ResourceController
+}
+
+// JobOpts wraps the options that can be passed to cgroups for the job
+// details at https://facebookmicrosites.github.io/cgroup2/docs/overview
+type JobOpts struct {
+	CPUWeight int32      // `cpu.weight`
+	IOWeight  int32      // `io.weight`
+	MemLimit  CgroupByte // `mem.high`
+}
+
+// JobStatus is an amalgamation of the useful status information available from the exec.Cmd struct of the job and it's underlying os.Process
+type JobStatus struct {
+	ID       string
+	PID      int
+	Running  bool
+	ExitCode int
+}
+
+func (status JobStatus) String() string {
+	return fmt.Sprintf(`Job Status
+	ID	%s
+	PID	%d
+	Running	%t
+	ExitCode %d`, status.ID, status.PID, status.Running, status.ExitCode)
+}
+
+// ResourceController defines the interface for implementing resource control of new processes
+// In cgroups this will be creating, editing and deleting files in /sys/fs/cgroup
+type ResourceController interface {
+	AddProcess(string, *exec.Cmd) error
+	CreateGroup(string) error
+	DeleteGroup(string) error
+	AddResourceControl(string, JobOpts) error
+}
+
+// NewOpts returns a configured JobOpts based on arguments
+func NewOpts(weight, ioLatency int32, memLimit CgroupByte) JobOpts {
+	return JobOpts{
```

Comment:
I actually prefer inline initialisation for these, as `NewOpts()` just hides what each value means at the call site.